packages/assets-controllers/src/MultichainAssetsController/MultichainAssetsController.ts

```diff
@@ -0,0 +1,430 @@
+ * the `anonymous` flag.
+ */
+const assetsControllerMetadata = {
+  metadata: {
+    persist: true,
+    anonymous: false,
+  },
+  allNonEvmTokens: {
+    persist: true,
+    anonymous: false,
+  },
+};
+
+// Define a temporary interface for the permission structure
+type AssetEndowment = {
+  'endowment:assets'?: {
+    scopes: CaipChainId[];
+  };
+};
+
+export class MultichainAssetsController extends BaseController<
+  typeof controllerName,
+  MultichainAssetsControllerState,
+  MultichainAssetsControllerMessenger
+> {
+  // Mapping of CAIP-2 Chain ID to Asset Snaps.
+  #snaps: Record<CaipChainId, Snap[]>;
+
+  constructor({
+    messenger,
+    state = {},
+  }: {
+    messenger: MultichainAssetsControllerMessenger;
+    state?: Partial<MultichainAssetsControllerState>;
+  }) {
+    super({
+      messenger,
+      name: controllerName,
+      metadata: assetsControllerMetadata,
+      state: {
+        ...getDefaultMultichainAssetsControllerState(),
+        ...state,
+      },
+    });
+
+    this.#snaps = {};
+
+    this.messagingSystem.subscribe(
+      'AccountsController:accountAdded',
+      async (account) => await this.#handleOnAccountAdded(account),
+    );
+    this.messagingSystem.subscribe(
+      'AccountsController:accountRemoved',
+      async (account) => await this.#handleOnAccountRemoved(account),
+    );
+  }
+
+  /**
+   * Checks for non-EVM accounts.
+   *
+   * @param account - The new account to be checked.
+   * @returns True if the account is a non-EVM account, false otherwise.
+   */
+  #isNonEvmAccount(account: InternalAccount): boolean {
+    return (
+      !isEvmAccountType(account.type) &&
+      // Non-EVM accounts are backed by a Snap for now
+      account.metadata.snap !== undefined
+    );
+  }
+
+  /**
+   * Handles changes when a new account has been added.
+   *
+   * @param account - The new account being added.
+   */
+  async #handleOnAccountAdded(account: InternalAccount): Promise<void> {
+    if (!this.#isNonEvmAccount(account)) {
+      // Nothing to do here for EVM accounts
+      return;
+    }
+
+    // Get assets list
+    if (account.metadata.snap) {
+      const assets = await this.#getAssetsList(
+        account.id,
+        account.metadata.snap.id,
+      );
+      const assetsWithoutMetadata: CaipAssetType[] = assets.filter(
+        (asset) => !this.state.metadata[asset],
+      );
+
+      // call the snap to get the metadata
+      if (assetsWithoutMetadata.length > 0) {
+        // check if for every asset in assetsWithoutMetadata there is a snap in snaps by chainId else call getAssetSnaps
+        if (
+          !assetsWithoutMetadata.every((asset: CaipAssetType) => {
+            const chainId = parseCaipAssetType(
+              asset as `${string}:${string}/${string}:${string}`,
+            );
+            console.log('this.snaps', this.#snaps);
+            return this.#snaps[chainId as CaipChainId]?.length > 0;
```

Comment:
We're also having this pattern when using this notation:

```ts
const [snap] = this.#snaps[chainId as CaipChainId];
```

So IMO, that would be great to have another method for this like:

```ts
#getAssetSnapFor(scope: CaipChainId): Snap {
  const [snap] = this.#snaps[scope];
  return snap; // Will be undefined if there's no Snaps candidate for this scope.
}
```

And this would just become:
```suggestion
            return Boolean(this.#getAssetSnapFor(chainId));
```

Or using `!== undefined` if you prefer.

Reply:
I updated the `getAssetSnapFor` slightly because if `this.#snaps = {}` we would get a type error trying to do
this.#snaps[scope]